ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PullQuerySingleNodeFunctionalTest.java

```diff
@@ -0,0 +1,346 @@
+    String stateDir = (String)REST_APP_0.getBaseConfig()
+        .get(KSQL_STREAMS_PREFIX + StreamsConfig.STATE_DIR_CONFIG);
+    clearDir(stateDir);
+
+    // Pause incoming kafka consumption
+    APP_SHUTOFFS_0.setKafkaPauseOffset(2);
+
+    LOG.info("Restarting the server " + host0.toString());
+    REST_APP_0.start();
+
+    waitForStreamsMetadataToInitialize(REST_APP_0, ImmutableList.of(host0), queryId);
+    waitForRemoteServerToChangeStatus(REST_APP_0, host0, HighAvailabilityTestUtil
+        .lagsReported(host0, Optional.of(2L), 5));
+
+    ClusterStatusResponse clusterStatusResponse = HighAvailabilityTestUtil
+        .sendClusterStatusRequest(REST_APP_0);
+    Pair<Long, Long> pair = getOffsets(host0, clusterStatusResponse.getClusterStatus());
+    assertThat(pair.left, is(2L));
+    assertThat(pair.right, is(5L));
+
+    final List<StreamedRow> sameRows = makePullQueryRequest(
+        REST_APP_0, sql, LAG_FILTER_3);
+
+    host = sameRows.get(1).getSourceHost().get();
+    assertThat(host.getHost(), is(host0.getHost()));
+    assertThat(host.getPort(), is(host0.getPort()));
+    assertThat(sameRows.get(1).getRow(), is(not(Optional.empty())));
+    // Still haven't gotten the update yet
+    assertThat(sameRows.get(1).getRow().get().values(), is(ImmutableList.of(KEY, 1)));
+
+    // Row not found!
+    final List<StreamedRow> headerOnly = makePullQueryRequest(
+        REST_APP_0, sqlKey3, LAG_FILTER_3);
+    assertThat(headerOnly.size(), is(1));
+
+    // Unpause incoming kafka consumption. We then expect active to catch back up.
+    APP_SHUTOFFS_0.setKafkaPauseOffset(-1);
+
+    waitForRemoteServerToChangeStatus(REST_APP_0, host0, HighAvailabilityTestUtil
+        .lagsReported(host0, Optional.of(5L), 5));
+
+    clusterStatusResponse = HighAvailabilityTestUtil
+        .sendClusterStatusRequest(REST_APP_0);
+    pair = getOffsets(host0, clusterStatusResponse.getClusterStatus());
+    assertThat(pair.left, is(5L));
+    assertThat(pair.right, is(5L));
+
+    final List<StreamedRow> updatedRows = makePullQueryRequest(
+        REST_APP_0, sqlKey3, LAG_FILTER_3);
+
+    // Now it is found!
+    host = updatedRows.get(1).getSourceHost().get();
+    assertThat(host.getHost(), is(host0.getHost()));
+    assertThat(host.getPort(), is(host0.getPort()));
+    assertThat(updatedRows.get(1).getRow(), is(not(Optional.empty())));
+    assertThat(updatedRows.get(1).getRow().get().values(), is(ImmutableList.of(KEY_3, 1)));
+  }
+
+  private static String extractQueryId(final String outputString) {
+    final java.util.regex.Matcher matcher = QUERY_ID_PATTERN.matcher(outputString);
+    assertThat("Could not find query id in: " + outputString, matcher.find());
+    return matcher.group(1);
+  }
+
+  private static void makeAdminRequest(TestKsqlRestApp restApp, final String sql) {
+    RestIntegrationTestUtil.makeKsqlRequest(restApp, sql, Optional.empty());
+  }
+
+  private static List<KsqlEntity> makeAdminRequestWithResponse(
```

Comment:
These are common methods with the PullQueryRoutingFunctionalTest. Maybe move them to the HATestUtil class so that they are in one place?

Reply:
Sure, sounds good.